### ISSUE TITLE
"This expression is not callable" with module: "node16" in tsconfig.json

### ISSUE THREAD

**Issue by karlhorky** [sebbo2002/ical-generator#469]
Hi @sebbo2002 👋 Thanks again for `ical-generator`, very useful!

The new module resolution option `node16` does not work with `ical-generator` unfortunately, failing with the `This expression is not callable` error:

```
This expression is not callable.
  Type 'typeof import("stackblitz:/node_modules/ical-generator/dist/index")' has no call signatures.
```

StackBlitz demo: https://stackblitz.com/edit/node-mmyw9w?file=tsconfig.json,package.json&file=tsconfig.json,index.ts

<img width="1200" alt="Screenshot 2023-03-09 at 22 07 20" src="https://user-images.githubusercontent.com/1935696/224158987-6212c06f-3dde-4f93-bceb-bde055e846e3.png">

The problem is the `export default` in [the published `index.d.ts`](https://unpkg.com/browse/ical-generator@3.6.1/dist/index.d.ts), as far as I understand, since `ical-generator` uses a CommonJS `module.exports =` export, this should actually instead be `export =`.

---

You can also see this on the tool [Are The Types Wrong](https://arethetypeswrong.github.io/) from @andrewbranch:

<img width="914" alt="Screenshot 2023-03-09 at 22 09 07" src="https://user-images.githubusercontent.com/1935696/224158543-9ec6650e-9efe-40df-86e2-03faf8a254e6.png">

**Comment by sebbo2002** [comment]
Thanks for this great bug report. Unfortunately, I don't see at first glance why the type definitions are being generated incorrectly and since I have a visitor this weekend, I probably won't be able to deal with it extensively until next week. I hope that's ok, otherwise pull requests are always appreciated of course. Thank you.

**Comment by karlhorky** [comment]
Also not 100% sure what `tsconfig.json` options cause this, but if I would have to guess, I would look at this `compilerOptions.target: 'es6'` - seems like an outdated module format. Also `module: 'none'` is interesting... 🤔

https://github.com/sebbo2002/ical-generator/blob/f105df90729bfa46e477922303b2ce1222a89aa4/tsconfig.json#L5-L10

**Comment by karlhorky** [comment]
Maybe `module: "commonjs"` is needed, as per [this Stack Overflow answer](https://stackoverflow.com/a/73800423/1268612)?

**Comment by sebbo2002** [comment]
Do you want to try it with the current develop version (`ical-generator@4.0.0-develop.1`)?

<img width="625" alt="Bildschirm­foto 2023-03-17 um 23 15 49" src="https://user-images.githubusercontent.com/812398/226062444-da7e4da1-7222-42d0-8946-ae3491e2a4a9.png">

**Comment by karlhorky** [comment]
Forked the StackBlitz demo to a new one, seems like it's working 🙌 

Thanks! Closing this.

Updated (working) StackBlitz demo: https://stackblitz.com/edit/node-5l65bv?file=tsconfig.json,package.json&file=tsconfig.json,index.ts

<img width="1305" alt="Screenshot 2023-03-19 at 19 21 43" src="https://user-images.githubusercontent.com/1935696/226198632-4881d883-d524-4fff-87e4-ebdff1c07a2e.png">

**Comment by karlhorky** [comment]
@sebbo2002 reopening this - `ical-generator@4.0.0-develop.2` is broken again, check it out on [Are The Types Wrong?](https://arethetypeswrong.github.io/)

<img width="856" alt="Screenshot 2023-03-24 at 09 50 09" src="https://user-images.githubusercontent.com/1935696/227470702-8b502f8e-5de7-492d-893c-a4b0ab902d24.png">

**Comment by sebbo2002** [comment]
Yes, I've seen when merged it in, but I don't understand the error yet and I haven't had time to look at it. The `.d.ts` file is the same, yes, but I only have one type definition file after running `tsup`. If I duplicate the file so that the filenames are different, everything is okay. But that can't possibly be the cause.

**Comment by karlhorky** [comment]
Haven't looked at your project configuration / publishing formats yet, but maybe you need something like this?

- https://github.com/tricinel/highlight-words/issues/30#issuecomment-1408321591

**Comment by karlhorky** [comment]
Yeah the `*.d.mts` files seemed counterintuitive to me too, but [Andrew Branch](https://github.com/andrewbranch) suggested them, so I think it's a good way to go

**Comment by karlhorky** [comment]
Sweet, how did you get the compiler to emit [the `dist/index.d.cts` file](https://unpkg.com/browse/ical-generator@4.0.0/dist/index.d.cts)?

Was it just this change to the `tsconfig.json` in https://github.com/sebbo2002/ical-generator/pull/462?

```diff
-        "module": "none",
-        "moduleResolution": "node",
+        "module": "nodenext",
+        "moduleResolution": "nodenext",
```

I'd like to learn how you did it to be able to recommend things for other packages too...

If you have any other changes that should be highlighted, happy to hear about those too!

**Comment by karlhorky** [comment]
Oh wait, maybe it's this manual `cp` that you did here? 🤔

```diff
-    "build": "tsc",
+    "build": "tsup && cp ./dist/index.d.ts ./dist/index.d.cts",
```

I thought there would have been a way for `tsc` to create these files automatically 😯

**Comment by marc-wittwer** [comment]
@sebbo2002  It seems that this currently does not work for  "node16" in tsconfig.json with Node v20.9.0 and v20.11.1
```
import ical from "ical-generator";

const cal = ical();

Error:
src/index.ts(13,15): error TS2349: This expression is not callable.
  Type 'typeof import("/project/workspace/node_modules/ical-generator/dist/index")' has no call signatures.
```

Reproduction example: https://codesandbox.io/p/devbox/simple-express-server-f7ldd?file=%2Fpackage.json


```
//tsconfig.json
 "compilerOptions": {
        "target": "ES2022",
        "module": "Node16" 
        "moduleResolution": "Node16",
```
___
### I looked into ways to fix it:
**Option 1:**
```
// Use the calendar item class directly:
import { ICalCalendar } from "ical-generator";
const calendar = new ICalCalendar();
console.log(calendar);

// This works without any problems
```

**Option 2:**
Add a named export to the ical-generator package for ical.

```diff
// src/index.ts in ical-generator
-  function ical(data?: ICalCalendarData): ICalCalendar {
+ export function ical(data?: ICalCalendarData): ICalCalendar {
    return new ICalCalendar(data);
}

export default ical;

// usage:
import { ical } from "ical-generator";
```
It could be that it is related to this: https://github.com/microsoft/TypeScript/issues/50175
